doc/web-socket.rst

```diff
@@ -0,0 +1,405 @@
+.. _web-socket:
+
+Web Socket
+==========
+
+Broker offers access to the publish/subscribe layer via WebSocket in order to
+make its :ref:`data <data-model>` accessible to third parties.
+
+WebSocket clients are treated as lightweight peers. Each Broker endpoint can be
+configured to act as a WebSocket server by either setting the environment
+variable ``BROKER_WEB_SOCKET_PORT`` or by setting ``broker.web-socket.port`` on
+the command line or in the configuration file. When running Zeek, scripts may
+also redefine ``Broker::web_socket_port`` to a valid TCP port.
+
+.. note::
+
+  Broker uses the same SSL parameters for native and WebSocket peers.
+
+JSON API v1
+-----------
+
+To access the JSON API, clients may connect to
+``wss://<host>:<port>/v1/events/json`` (SSL enabled, default) or
+``ws://<host>:<port>/v1/events/json`` (SSL disabled). On this WebSocket
+endpoint, Broker allows JSON-formatted text messages only.
+
+Handshake
+~~~~~~~~~
+
+The broker endpoint expects a JSON array of strings as the first message. This
+array encodes the subscriptions as a list of topic prefixes that the client
+subscribes to. Clients that only publish data must send an empty JSON array.
+
+After receiving the subscriptions, the Broker endpoint sends a single ACK
+message:
+
+.. code-block:: json
+
+  {
+    "type": "ack",
+    "endpoint": "<uuid>",
+    "version": "<broker-version>"
+  }
+
+In this message, ``<uuid>`` is the unique endpoint ID of the WebSocket server
+and ``<broker-version>`` is a string representation of the ``libbroker``
+version, i.e., the result of ``broker::version::string()``. For example:
+
+.. code-block:: json
+
+  {
+    "type": "ack",
+    "endpoint": "925c9110-5b87-57d9-9d80-b65568e87a44",
+    "version": "2.2.0-22"
+  }
+
+Protocol
+~~~~~~~~
+
+After the handshake, the WebSocket client may only send `Data Messages`_. The
+Broker endpoint converts every message to its native representation and
+publishes it.
+
+The WebSocket server may send `Event Messages`_ (whenever a data message matches
+the subscriptions of the client) and `Error Messages_` to the client.
+
+Data Representation
+~~~~~~~~~~~~~~~~~~~
+
+Broker uses a recursive data type to represent its values (see
+:ref:`Data Model <data-model>`). This data model does not map to JSON-native
+types without ambiguity, e.g., because Broker distinguishes between signed and
+unsigned number types.
+
+In JSON, we represent each value as a JSON object with two keys: ``@data-type``
+and ``data``. The former identifies one of Broker's data types (see below) and
+denotes how Broker parses the ``data`` field.
+
+None
+****
+
+There is only exactly one valid input for encoding a ``none``:
+
+.. code-block:: json
+
+  {
+    "@data-type": "none",
+    "data": {}
+  }
+
+
+Boolean
+*******
+
+The type ``boolean`` can take on exactly two values and maps to the native JSON
+boolean type:
+
+.. code-block:: json
+
+  {
+    "@data-type": "boolean",
```

Comment:
Why's this prefixed with `@`, whereas other keys aren't? (Just curious).

Reply:
The idea is to flag meta fields with an `@` since they are more like annotations than actual fields. In this case, this is an annotation for the `data` field that tells the parser what type to use.